File: 2017/Day_01/Python/solution.py
"""Solution to AoC 2017 Day 1 - Inverse Captcha"""

import re

def input_only_one_line(file: str):
    """Puzzle input is just one line."""
    with open(file, 'r') as input_file:
        return input_file.readline()


def detect_double_numbers(number: str) -> int:
    """Detect double numbers and sum them."""
    regex = re.compile(r'(\d)(?=\1)')
    double_numbers = re.findall(regex, number)
    number_sum = 0
    for the_number in double_numbers:
        number_sum += int(the_number)
    if number[0] == number[-1]:
        number_sum += int(number[0])
    return number_sum


if __name__ == "__main__":
    input_number = input_only_one_line("../input.txt")
    part_one = detect_double_numbers(input_number)
    print(f"The part 1 reverse captcha is {part_one}")


# 942 is too low

File: 2017/Day_01/Perl/solution.pl
#! /usr/bin/perl

use v5.20;

open (CAPTCHA, "../input.txt") || die "No captcha!";

my $captcha = <CAPTCHA>;

my @double_numbers = $captcha =~ /(\d)(?=\1)/g;

my $number_sum = 0;

foreach my $number (@double_numbers){

    $number_sum += $number;

}

my @numberlist = split('', $captcha);

if (@numberlist[0] == @numberlist[-1]){

    $number_sum += @numberlist[0];

}

say "Part 1 captcha is $number_sum";

$number_sum = 0;

my $lookahead = (scalar @numberlist) / 2;

my $index = 0;

foreach my $number (@numberlist){

    my $check = ($index + $lookahead) % (scalar @numberlist);
    
    if ($number == @numberlist[$check]){
    
        $number_sum += $number;
    
    }
    
    $index++;

}

say "Part 2 captcha is $number_sum";
